## Problem #133: Repunit Nonfactors
[Link to Problem](https://projecteuler.net/problem=133)

### Problem Description

A number consisting entirely of ones is called a repunit. We shall define $R(k)$ to be a repunit of length $k$; for example, $R(6) = 111111$.

Let us consider repunits of the form $R(10^n)$.

Although $R(10)$, $R(100)$, or $R(1000)$ are not divisible by $17$, $R(10000)$ is divisible by $17$. Yet there is no value of $n$ for which $R(10^n)$ will divide by $19$. In fact, it is remarkable that $11$, $17$, $41$, and $73$ are the only four primes below one-hundred that can  be a factor of $R(10^n)$.

Find the sum of all the primes below one-hundred thousand that will never be a factor of $R(10^n)$.

### Approach

First of all, $R(x) = \frac{10^x - 1}{9}$.

Then if we use some modular arithmetic:

$$

R(10^n) = \frac{10^{10^n} - 1}{9} \\[5pt]

\frac{10^{10^n} - 1}{9} \equiv 0 \pmod{p} \\[5pt]

\gcd(9, p) = 1, \; \forall \; p \geq 100 \\[5pt]

\Rightarrow  10^{10^n} \equiv 1 \pmod{p} \\[5pt]

$$

Let's take a step back, if $10^x \equiv 1 \pmod{p}$, then $x$ must divide $\varphi(p) = p - 1$.

Therefore if we find the minimum value for $x$, and this $x$ is of the form $2^{e_1}5^{e_2}$, we can safely assume that there exists a big enough $n$ for which $R(X)$ is divisible by $p$.

In [4]:
limit = 100000

sieve = [True] * limit
sieve[0] = sieve[1] = False

for i in range(2, limit):
    if sieve[i]:
        for j in range(i * i, limit, i):
            sieve[j] = False

can_be_factor = 0

for p in range(10, limit):
    if sieve[p]:
        x = p - 1
        for d in range(1, int((p - 1) ** 0.5) + 1):
            if (p - 1) % d == 0:
                if pow(10, d, p) == 1:
                    x = min(x, d)
                if pow(10, (p - 1) // d, p) == 1:
                    x = min(x, (p - 1) // d)
        while x % 2 == 0:
            x //= 2
        while x % 5 == 0:
            x //= 5
        if x == 1:
            can_be_factor += p

total = sum([index for index, value in enumerate(sieve) if value])

print(f"Answer: {total - can_be_factor}")

Answer: 453647705


###### Result: **453647705** | Execution time: ~0s

### Complexity analysis

Time complexity: $O(limit\sqrt{limit})$

##### Tags: #number-theory, #modular-arithmetic, #primes